newdle/api.py

```diff
@@ -323,36 +324,68 @@ def create_newdle(title, duration, timezone, timeslots, participants, private, n
             ),
         },
     )
+    db.session.commit()
     return NewdleSchema().jsonify(newdle)
 
 
-@api.route('/newdle/<code>')
-@allow_anonymous
-def get_newdle(code):
-    newdle = Newdle.query.filter_by(code=code).first_or_404(
-        'Specified newdle does not exist'
-    )
-    if newdle.deleted:
-        return DeletedNewdleSchema().jsonify(newdle)
-    return RestrictedNewdleSchema().jsonify(newdle)
-
-
 @api.route('/newdle/<code>', methods=('PATCH',))
-@use_args(UpdateNewdleSchema(), locations=('json',))
+@use_args(NewNewdleSchema(partial=True), locations=('json',))
 def update_newdle(args, code):
     newdle = Newdle.query.filter_by(code=code).first_or_404(
         'Specified newdle does not exist'
     )
-    if newdle.creator_uid != g.user['uid']:
+    if g.user is None or newdle.creator_uid != g.user['uid']:
         raise Forbidden
+    new_participants = []
+    if 'participants' in args:
+        participants = args.pop('participants')
+        # Filter the new participants to be created (left diff)
+        new_participants = [
+            Participant(**_p)
+            for _p in participants
+            if not any(_p['auth_uid'] == p.auth_uid for p in newdle.participants)
+        ]
+        # Filter the existing participants so we don't reset them (intersection)
+        newdle.participants = {
+            p
+            for p in newdle.participants
+            if any(p.auth_uid == _p['auth_uid'] for _p in participants)
+        }
+        newdle.participants.update(new_participants)
     for key, value in args.items():
         setattr(newdle, key, value)
     if args:
         newdle.update_lastmod()
+    db.session.flush()
+    notify_newdle_participants(
+        newdle,
+        f'Invitation: {newdle.title}',
+        'invitation_email.txt',
+        'invitation_email.html',
+        lambda p: {
+            'creator': newdle.creator_name,
+            'title': newdle.title,
+            'answer_link': url_for(
+                'newdle', code=newdle.code, participant_code=p.code, _external=True
+            ),
+        },
+        participants=new_participants,
+    )
```

Comment:
I guess this is very similar to what we have when creating a newdle? If yes, maybe we could add a `send_invitation_emails(newdle, participants=None)` util and call this in both places?

Reply:
Should we include this together in the `notifications.py` file or create new util and move some of the email method-objects there?

Reply:
`notifications.py` seems like a good place